servicetalk-serializer-utils/src/main/java/io/servicetalk/serializer/utils/StringCharsetSerializer.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.serializer.utils;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.serializer.api.SerializerDeserializer;
+
+import java.nio.charset.Charset;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Serialize/deserialize {@link String}s encoded with a {@link Charset}.
+ * @see StringAsciiSerializer
+ * @see StringUtf8Serializer
+ */
+public final class StringCharsetSerializer implements SerializerDeserializer<String> {
+    private static final Map<Charset, Integer> MAX_BYTES_PER_CHAR_MAP = new HashMap<>();
+    static {
```

Comment:
There are often hundreds of mostly garbage charsets. Perhaps computeIfAbsent?

Reply:
can you clarify the benefit of `computeIfAbsent`? are you implying there will be `Charset` whose hash/equals collide and the first one should win?

Reply:
Ah, commented on the wrong line. 

The problem is that `Charset.availableCharsets().values()` may return hundreds of charsets, mostly goofy Windows ones, obscure CJK charsets and oddballs like EBCDIC. The OpenJDK default is about 125 charsets, SAP and IBM in particular include many more in their JDKs (400+ if I remember correctly). Not sure about other vendors. With `computeIf Absent(charset, aCharset -> (int) charset.newEncoder().maxBytesPerChar());` you'll only compute the charsets you need and there is no need for the `getOrDefault(charset, 1)` should you encounter a charset not in `Charset.availableCharsets()`, which is possible.

Reply:
Lets defer this for later if more exotic Charsets are used in practice and if we dynamically modify the map in the constructor we have to worry about concurrent access which I'm assuming this "missing charset" case will be very rare (most common usage is in `StandardCharset` which I assume would also be in `availableCharsets`).

Reply:
Yes, `StandardCharset` will be in `availableCharsets`. I am still worried about the cost of computing the result for all available charsets when most (probably 95%+) will never be used and resulting impact on startup time. I noticed that another class, `FormUrlEncodedSerializer` also builds a map for all charsets. A possible alternative is to continue to use HashMap, preload it with the StandardCharsets, and when another charset is encountered, copy the map, add the new charset and replace the entire map. Lighter weight than a full `ConcurrentMap` and only a volatile read more expensive on average.

Reply:
I will add a utility targeting standardard charset (I don't know of an existing collection which makes these available from the JDK). Since no COW map exists in the JDK I'll just stick to read only map for now and we can enhance later if necessary.